### PR TITLE
Improve diagnostics for unterminated nested block comment

### LINKED ISSUE #95283: Better diagnostics for accidentally nested block comments
Given:
```rust
fn main ()
{
    /*
    //bland
    //weconwec'iwe
    
    //wec'oinwec'oinwec
    
    //wec
    //wecponjwec
    
    
    //wecponwec
    //let foo: &u32 = panic!();
    //*foo = 12
    //help out
    //more stuff
    */
}
```

The compiler says:
```
error[[E0758]](https://doc.rust-lang.org/stable/error-index.html#E0758): unterminated block comment
  [--> src/main.rs:3:5
](https://play.rust-lang.org/#)   |
3  | /     /*
4  | |     //bland
5  | |     //weconwec'iwe
6  | |     
...  |
18 | |     */
19 | | }
   | |_^

For more information about this error, try `rustc --explain E0758`.
error: could not compile `playground` due to previous error
```

This initially confused me quite a bit because I couldn't figure out why Rust wasn't seeing block comment terminator. Eventually, I realized that the line `//*foo = 12` changed from being a commented out line, to starting a nested block comment. Since block comments nest in Rust, I'm used to commenting out blocks without worrying about what's inside of them which made this behaviour surprising.

I'm not sure how practical it is, but it would be cool if we could check for comments like `//*foo = 12` in block comments to suggest them as the source of the problem.

### PR DESCRIPTION
**Pull request by yue4u**
close #95283

(This is my first time try to messing around with rust compiler and might get a lot of things wrong... :bow: )

### CODE REVIEW

**Comment by petrochenkov** [comment]
The text of the comment should be available in `fn cook_lexer_token`, so I suggest searching for the nested `/*` in it only if we are actually going to report the "unterminated block (doc-)comment" error.

Also all this logic searching for `/*` and adding the labels can be moved to a separate function and placed somewhere together with other error-reporting functions.

File: compiler/rustc_parse/src/lexer/mod.rs [hunk 1]
```diff
@@ -553,6 +544,58 @@ impl<'a> StringReader<'a> {
...
+            self.mk_sp(start, last_bpos),
+            msg,
+            error_code!(E0758),
+        );
+        let mut nested_block_comment_open_idxs = vec![];
+        let mut last_nested_block_comment_idxs = None;
+        let mut content_chars = self.str_from(start).char_indices();
+
+        if let Some((_, mut last_char)) = content_chars.next() {
+            while let Some((idx, c)) = content_chars.next() {
```
Comment (rainy-me) [inline compiler/rustc_parse/src/lexer/mod.rs:563]: Maybe using zip or windows(2) will be much more readable, but I didn't found any usage of both under rustc_parse so not sure if it's ok to use them.

**Review by yue4u — COMMENTED** [review]
Hi @petrochenkov , thanks for the review. I updated this pr and does this look better?

File: compiler/rustc_parse/src/lexer/mod.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_parse/src/lexer/mod.rs:563]: It's ok to use any standard library features in this crate if they improve the code.

Comment (rainy-me) [reply compiler/rustc_parse/src/lexer/mod.rs]: I decided to change to use `peekable` instead as `itertools` is not standard library and zip seems need clone 1b7008dc77e25049b04e5c3e31aecf4de00803e7

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**